chapters/mapping_data_to_shaders.md

```diff
@@ -0,0 +1,308 @@
+```
+#version 450
+layout (constant_id = 0) const float myColor = 1.0;
+layout(location = 0) out vec4 outColor;
+
+void main() {
+    outColor = vec4(myColor);
+}
+```
+
+Resulting SPIR-V assembly:
+```
+             Decorate 9(outColor) Location 0
+             Decorate 10(myColor) SpecId 0
+
+// 0x3f800000 as decimal which is 1.0 for a 32 bit float
+10(myColor): 6(float) SpecConstant 1065353216
+```
+
+With specialization constants, the value is still a constant inside the shader, but for example, if another `VkPipeline` uses the same shader, but wants to set the `myColor` value to `0.5f`, it is possible to do so at runtime.
+```
+struct myData {
+    float myColor = 1.0f;
+} myData;
+
+VkSpecializationMapEntry mapEntry = {};
+mapEntry.constantID = 0; // matches constant_id in GLSL and SpecId in SPIR-V
+mapEntry.offset     = 0;
+mapEntry.size       = sizeof(float);
+
+VkSpecializationInfo specializationInfo = {};
+specializationInfo.mapEntryCount = 1;
+specializationInfo.pMapEntries   = &mapEntry;
+specializationInfo.dataSize      = sizeof(myData);
+specializationInfo.pData         = &myData;
+
+VkGraphicsPipelineCreateInfo pipelinenfo = {};
+pipelinenfo.pStages[fragIndex].pSpecializationInfo = &specializationInfo;
+
+// Create first pipeline with myColor as 1.0
+vkCreateGraphicsPipelines(&pipelineInfo);
+
+// Create second pipeline with same shader, but sets different value
+myData.myColor = 0.5f;
+vkCreateGraphicsPipelines(&pipelineInfo);
+```
+
+The second `VkPipeline` shader disassembled has the new constant value for `myColor` of `0.5f`.
+
+### 3 Types of Specialization Constants Usages
+
+The typical use cases of specialization constant can be best grouped into three different usages.
+
+1. Toggling features
+    - The support a feature in Vulkan isn't known until runtime. This usage of specialization constant is to prevent writing two separate shaders, but instead embedded a constant runtime decision.
+    -
+    ```
+    #version 450
+    #extension GL_KHR_shader_subgroup_vote : enable
+    layout (constant_id = 0) const bool voteEnabled = true;
+
+    void main() {
+        // ...
+        if (voteEnabled == true) {
+            // Only valid if Vulkan exposes vote subgroup operations
+            myResults = subgroupAll(true);
+        } else {
+            myResults = false;
+        }
+        // ...
+    }
+    ```
+2. Improving backend optimizations
+    - The "backend" here refers the the implementation's compiler that takes the resulting SPIR-V and lowers it down to some ISA to run on the device.
+    - Constant values allow a set of optimizations such as [constant folding](https://en.wikipedia.org/wiki/Constant_folding), [dead code elimination](https://en.wikipedia.org/wiki/Dead_code_elimination), etc. to occur.
+    - While divergence gives poor performance in shader code, if the compiler knows from the SPIR-V specialization constant that only one branch is possible, it can easily eliminate the branch for you.
```

Comment:
This isn’t strictly true, and is something of an artefact of days past. The cases where specialisation constants could otherwise be used would result in uniform branching - which is extremely low cost. Only non-uniform branches are actively expensive, and you can’t replace those with specialisation constants.

In cases where specialising would result in a high number of pipeline objects, it’s often a better idea to perform dynamic uniform branching to avoid the compilation cost.